rfd/0213-relay-a-lightweight-proxy.md

```diff
@@ -0,0 +1,66 @@
+
+The discovery data will be occasionally re-fetched as a safeguard against broken relays resulting in isolation of an agent, but it's expected that any change to the available relays as a result of rollouts and scaling will be quickly propagated to all relays through the usual event fanout, and will then be broadcast to all the connected agents (similarly to how the regular reverse tunnel works). It is not practical to have all relay heartbeats replicated to individual agents through their own event watchers, since heartbeat-like resources cached by large volumes of agents are a performance problem for the event fanout as well as a noticeable load on the network. The periodic fetching of the discovery data is also how agents will come to learn about changes to the agent connection count, if any.
+
+It should be possible for the relay reverse tunnels to support all services immediately, since no changes should be necessary for the connections tunneled from the Teleport control plane to the agents.
+
+### Client behavior
+
+The relay group address can be specified at `tsh login` time, and it can be explicitly specified and/or disabled as part of the invocation of `tsh ssh`, `tsh proxy ssh`, `tsh kube login` or `tsh proxy kube`, with either a command line option or an environment variable; as usual, the command line option overrides the envvar which overrides the configuration stored at login time. Access through machine ID will need explicit configuration as part of the "output" or "service" in use, and any need to rapidly switch between relay and non-relay routing can be served by creating two copies of a given output or service. At least in the first implementation of the client tooling side of this feature, there will be no support for automatic switching or fallback between relay and non-relay connection routing, as well as no safeguards against connecting through the broader control plane even if a relay could've been used.
+
+Specifying a relay address at login time in Teleport Connect is left for future work, but a possible way to allow that could be an "Advanced..." button in the "Enter cluster address" modal, or some UI element to change settings for the active cluster after logging in. Editing the profile configuration file is also an option (albeit one that we don't necessarily want to encourage or support), as well as using `tsh ssh --relay <relayaddr> <user>@<host>` in a terminal tab instead of choosing a server to connect to in the list of resources.
+
+All the relays of the group will serve client connections in the same way, so clients only ever need to interact directly with the single load balancer address - this is a requirement to support Kubernetes clients without a local proxy - and will do so depending on the required protocol.
```

Comment:
That suggests Relays will support similar API surface to Proxy, feels contradictory to the earlier section about agent connectivity e.g. when joining. Can you clarify?

Reply:
Reworded to clarify.